java/ql/src/experimental/Security/CWE/CWE-489/ServletMain.ql

```diff
@@ -0,0 +1,51 @@
+/**
+ * @name Main Method in Servlet
+ * @description Jave EE web applications with a main method.
+ * @kind problem
+ * @id java/main-method-in-servlet
+ * @tags security
+ *       external/cwe-489
+ */
+
+import java
+import semmle.code.java.frameworks.Servlets
+
+/** The java type `javax.servlet.Filter` */
+class ServletFilterClass extends Class {
+  ServletFilterClass() { this.getASupertype*().hasQualifiedName("javax.servlet", "Filter") }
+}
+
+/** Listener class in the package `javax.servlet` and `javax.servlet.http` */
+class ServletListenerClass extends Class {
+  // Various listener classes of Java EE such as ServletContextListener. They all have a name ending with the word "Listener".
+  ServletListenerClass() {
+    this.getASupertype*()
+        .getQualifiedName()
+        .regexpMatch([
+            "javax\\.servlet\\.[a-zA-Z]+Listener", "javax\\.servlet\\.http\\.[a-zA-Z]+Listener"
+          ])
+  }
+}
+
+/** The `main` method in `Servlet`. */
+class ServletMainMethod extends Method {
+  ServletMainMethod() {
+    (
+      this.getDeclaringType() instanceof ServletClass or
+      this.getDeclaringType() instanceof ServletFilterClass or
+      this.getDeclaringType() instanceof ServletListenerClass
+    ) and
+    this.hasName("main") and
+    this.isStatic() and
+    this.getReturnType() instanceof VoidType and
+    this.isPublic() and
+    this.getNumberOfParameters() = 1 and
+    this.getParameter(0).getType() instanceof Array and
+    not this.getDeclaringType().getName().toLowerCase().matches("%test%") and // Simple check to exclude test classes to reduce FPs
```

Comment:
You can detect test classes a more accurately by this: https://github.com/github/codeql/pull/2628.

Reply:
@tamasvajk Thanks for reviewing this PR. For test classes detected by this query, they extends <code>Servlet</code> or <code>EnterpriseBean</code> instead of unit testing classes. Could you provide detailed explanation on how to detect test classes through the mechanism in [#2628](https://github.com/github/codeql/pull/2628)?

Reply:
Instead of `not this.getDeclaringType().getName().toLowerCase().matches("%test%")` you could have `not this.getDeclaringType() instanceof TestClass`.

Reply:
The change only helps if a class extends JUnit or TestNG. It won't help for a class named <code>ServletMainTest</code>, which is what the query tries to detect. Do I still miss something?

Reply:
It also identifies if there are methods in the class with test annotations. But maybe you're right that it's best to identify test servlets by name. In any case, I triggered two queries on lgtm.com, just to check:
* https://lgtm.com/query/59195713026144928/
* https://lgtm.com/query/5142109307273318911/

They seem to find the same issues. `/src/test/java` in the path seems to be a very effective filter :-) So let's leave this as it is.